### PR TITLE
Matrix-Free Linear Operator

### PR DESCRIPTION
**Pull request by e-moral-sanchez**
This PR provides a `MatrixFreeLinearOperator` class which is a subclass of the abstract class `LinearOperator`.

The new class allows to create a general matrix-free linear operator. The constructor only requires the domain, codomain and a callable `dot` method.

Notes:
- the provided `dot` method may or may not take an 'out' argument
- a `transpose_dot` method may also be provided (mandatory to instantiate the transpose() linear operator)
- unit tests have been added

Additional change: 
- in this PR we stop passing the deprecated `tol` argument in calls to scipy's minres. This forces us to use scipy >= 1.14 and hence python >= 3.9
- we also added a convergence verification in minres linear solver to avoid iterating when the initial solution is good enough (which may cause a division-by-zero error)

### CODE REVIEW

**Comment by campospinto** [comment]
This is useful I think, thanks! 

As a test, you can check the behaviour of a "fake" matrix-free operator: an operator created with a dot product involving a Stencil Matrix should act exactly like the matrix itself.

You can also check that an iterative solver (like cg) can run with an object of this new class.

**Comment by campospinto** [comment]
I see that some tests are failing because of `minres()` getting an unexpected keyword argument `tol`. Is this related to this new class?

**Review by jowezarek — APPROVED** [review]

**Comment by e-moral-sanchez** [comment]
The new version `1.14.0`  of SciPy removed the `tol` argument in `minres`: https://docs.scipy.org/doc/scipy-1.13.1/reference/generated/scipy.sparse.linalg.minres.html 

This also affects PR https://github.com/campospinto/psydac_dev/pull/6 . 

I will fix this in this PR.

File: psydac/linalg/basic.py [hunk 1]
```diff
@@ -1111,3 +1113,52 @@ def solve(self, rhs, out=None):
     @property
     def T(self):
         return self.transpose()
+
+#===============================================================================
+class MatrixFreeLinearOperator(LinearOperator):
+    """
+    General operator acting between two vector spaces V and W. It only requires a callable dot method.
+
```
Comment (campospinto) [inline psydac/linalg/basic.py:1148]: an example of instantiation would be useful

**Comment by campospinto** [comment]
> This PR provides the class `MatrixFreeLinearOperator`, which is a subclass of the abstract class `LinearOperator`.
> 
> The class `MatrixFreeLinearOperator` allows to create a general matrix-free linear operator. The constructor only requires the domain, codomain and a callable `dot` method.
> 
> TODO: add simple unit tests (ideas are welcome!)

I am just adding a few unit tests

**Comment by campospinto** [comment]
I am also adding some features to handle the case where the given dot method has not optional 'out' argument, and the possibility (not mandatory) to pass a 'transpose_dot' method

**Pull request merged**